python/ray/util/client/common.py

```diff
@@ -285,11 +285,20 @@ class ClientActorHandle(ClientStub):
 
     def __init__(self, actor_ref: ClientActorRef):
         self.actor_ref = actor_ref
+        self._dir: Optional[List[str]] = None
 
     def __del__(self) -> None:
         if ray.is_connected():
             ray.call_release(self.actor_ref.id)
 
+    def __dir__(self) -> List[str]:
+        if self._dir:
+            return self._dir
+        if ray.is_connected():
+            self._dir = ray.get(ray.remote(lambda x: dir(x)).remote(self))
```

Comment:
Launching a remote function within a tab completion seems wrong to me (imagine if it blocks due to unavailable resources or something).

Looking at how this is done in the `ActorHandle` class, it seems that the relevant method info is passed into the `ActorHandle` constructor, so we don't need to fetch it on the fly. Not sure if we have something equivalent here, but we eventually will (probably) in order to do static type checking for actor methods.

https://github.com/ray-project/ray/blob/45d2331d5a6062e894725024e2a608830ac0b7ce/python/ray/actor.py#L750-L918